### PR TITLE
Fix issue with Envoy not reference counting across scopes under not-hot restart

### PR DESCRIPTION
**Pull request by ambuc**
Signed-off-by: James Buckland <jbuckland@google.com>

*title*: Fixes issue with Envoy not reference counting stats across scopes under not-hot restart.

*Description*: Addresses issue https://github.com/envoyproxy/envoy/issues/2453, continuing draft work in https://github.com/ambuc/envoy/pull/1.  Summary of changes:
 - `HotRestartImpl` no longer is a child class of `RawStatDataAllocator`. It now has a member variable allocator, which is a more general-purpose `BlockRawStatDataAllocator`. This `BlockRawStatDataAllocator` accepts shared- or not- memory, and accepts / uses a mutex internally.
 - We now use this `BlockRawStatDataAllocator` in both `HotRestartImpl` and `HotRestartNopImpl`, which implements the reference counting by means of a `RawStatDataSet`. This leaves `HotRestartImpl` functionally unchanged, and extends that functionality to the no-op case.
 - This means `HotRestartNopImpl` needs to know a few `Options` and `BlockMemoryHashSetOptions` for creating its stat set, which means a small change to `main_common.cc` to pass in options to both kinds of restarters.
 - Associated changes to mocks / test suites to test this reference counting behaviors in both kinds of restart scenarios.

This also justifies the renaming changes in https://github.com/envoyproxy/envoy/pull/3150.

*Risk Level*: Medium.

*Testing*: Changed `stats_impl_test`, `thread_local_store_test`, `hot_restart_impl_test`, and `hot_restart_nop_impl_test` in large ways, and made minor updates to the server mocks. Added unit tests for `BlockRawStatDataAllocator` and the hot restart / no-op hot restart modes. Passes `bazel test test/...`.

*Docs Changes*: N/A

*Release Notes*: This is user-facing in that non-hot restart stat allocation now resolves namespace properly, but no effect on user configs.

Fixes: https://github.com/envoyproxy/envoy/issues/2453

### CODE REVIEW

**Comment by mattklein123** [comment]
@jmarantz do you mind taking a first pass here since you are familiar with these code paths?

**Comment by ggreenway** [comment]
The DCO-bot is having issues right now; there's probably nothing you can do at this time, so ignore it until we've sorted that out.

**Comment by jmarantz** [comment]
Thanks for the heads up @ggreenway -- James ignore my DCO comment :)

**Comment by ggreenway** [comment]
DCO-bot *should* be working now. It should run on the next commit pushed (or you can test it by pushing an empty commit to this PR). The previous failure does not indicate that there's anything wrong with the DCO-sign-off in this PR.

File: test/common/stats/stats_impl_test.cc [hunk 1]
```diff
@@ -468,5 +473,149 @@ TEST(TagProducerTest, CheckConstructor) {
...
+  auto foo_1 = allocator_->alloc("foo");
+  auto foo_2 = allocator_->alloc("foo");
+  EXPECT_EQ(foo_1, foo_2);
+  auto bar_1 = allocator_->alloc("bar");
+  EXPECT_NE(foo_1, bar_1);
+  EXPECT_NE(foo_2, bar_1);
+}
+
+TEST_F(BlockRawStatDataAllocatorTest, allocFail) {
+  EXPECT_CALL(options_, maxStats()).WillRepeatedly(Return(2));
```
[the remark this answers is not in the document]
Comment (mrice32) [reply test/common/stats/stats_impl_test.cc]: I think the calls it's expecting are made during `setup()`, right?

File: source/extensions/filters/http/dynamo/dynamo_utility.h [hunk 2]
```diff
@@ -9,6 +9,15 @@ namespace Dynamo {
 
 class Utility {
 public:
+  // returns example max name length, 127. This number is easily mutable in the
+  // outside world, and depending on it makes this test fragile. Thus, the examples
```
Comment (jmarantz) [inline source/extensions/filters/http/dynamo/dynamo_utility.h]: This doesn't look like a test utility; this looks like production code.

**Comment by mattklein123** [comment]
Can I ask a general question about this change: Was a much simpler solution considered? Basically in the heap allocator case, just maintain an unordered map with a reference count? That would be like 10 lines of code and I think would do the trick? Is there any benefit to the more complex approach?

**Comment by jmarantz** [comment]
TBH I had the same question but assumed that this done this way to share more code.  But the size of this PR indicates that maybe we'd be better off just using an unordered_set.

**Comment by mrice32** [comment]
I'd also like to add that the test failures that we've been seeing are related to test infrastructure issues (like preexisting static variables leaking across tests and cross platform types) rather than functional problems.  

I think using the block allocator has the following advantages: 
1.  It's a single implementation in both hot and !hot restart.  We don't currently run much CI for !hot restart outside of the mac tests, so reducing the differences makes our integration testing more robust and gives us bug-for-bug parity.
2.  Smaller concern, but I think having a single implementation reduces long term maintenance costs and will ultimately be fewer lines of total code.

I think the only disadvantage is that it will create limitations for the !hot restart case that aren't explicitly necessary.  So users who want to have 100,000 clusters will need to manually tune the size of their stat allocation despite the fact that it's on the heap.

That being said, I'm not too opinionated on which direction we go since both have merits.

**Comment by mattklein123** [comment]
I haven't really looked at the change, but I see a hundreds of line diff for something that could be trivially accomplished with a small amount of code, even if the implementations end up being different. This is also in an area that has a super high regression risk. My preference in general given this tradeoff is always going to be the trivial approach. But, if you all feel that sharing the code is the better way forward I won't fight it.

File: source/extensions/filters/http/dynamo/dynamo_utility.h (hunk 2 above)
Comment (ambuc) [reply source/extensions/filters/http/dynamo/dynamo_utility.h]: You're right. I messed up here -- thinking it was a utility which existed solely for `dynamo_utility_test.cc`. I think the thing breaking this test is a leaky stat

**Comment by ambuc** [comment]
Closed in favor of https://github.com/envoyproxy/envoy/pull/3212 as a simpler solution.

**Pull request closed without merging**